File: javascript/lasagna-master/lasagna-master.js
/// <reference path="./global.d.ts" />
// @ts-check

export function cookingStatus(time) {
  switch (time) {
    case 0:
      return "Lasagna is done.";
    case undefined:
      return "You forgot to set the timer.";
    default:
      return "Not done, please wait.";
  }
}

export function preparationTime(layers, time = 2) {
  return layers.length * time;
}

export function quantities(layers) {
  const noodlesQuantity = 50;
  const sauceQuantity = 0.2;
  let noodles = 0;
  let sauce = 0;

  layers.forEach((layer) => {
    noodles += layer === "noodles" ? noodlesQuantity : 0;
    sauce += layer === "sauce" ? sauceQuantity : 0;
  });

  return {
    noodles,
    sauce,
  };
}

export function addSecretIngredient(friendsList, myList) {
  myList.push(friendsList[friendsList.length - 1]);
}

export function scaleRecipe(recipe, portions) {
  const defaultPortions = 2;
  const factor = portions / defaultPortions;
  const newRecipe = {};

  for (let ingredient in recipe) {
    newRecipe[ingredient] = recipe[ingredient] * factor;
  }

  return newRecipe;
}

File: go/lasagna-master/lasagna_master.go
package lasagna

const (
	defaultPreparationTimePerLayer = 2
	noodleGramsPerLayer            = 50
	sauceLitersPerLayer            = 0.2
	defaultServingPortions         = 2
)

func PreparationTime(layers []string, minutes int) int {
	if minutes <= 0 {
		minutes = defaultPreparationTimePerLayer
	}

	return len(layers) * minutes
}

func Quantities(layers []string) (noodleGrams int, sauceLiters float64) {
	for _, layer := range layers {
		switch layer {
		case "noodles":
			noodleGrams += noodleGramsPerLayer
		case "sauce":
			sauceLiters += sauceLitersPerLayer
		}
	}

	return noodleGrams, sauceLiters
}

func AddSecretIngredient(friendsList, myList []string) {
	secretIngredient := friendsList[len(friendsList)-1]
	myList[len(myList)-1] = secretIngredient
}

func ScaleRecipe(quantities []float64, amount int) []float64 {
	scaledQuantities := make([]float64, len(quantities))
	multiplier := float64(amount) / defaultServingPortions

	for i := 0; i < len(quantities); i++ {
		scaledQuantities[i] = quantities[i] * multiplier
	}

	return scaledQuantities
}
